src/Microsoft.ML.AutoML/Experiment/Experiment.cs

```diff
@@ -49,23 +53,55 @@ internal class Experiment<TRunDetail, TMetrics> where TRunDetail : RunDetail
             _datasetColumnInfo = datasetColumnInfo;
             _runner = runner;
             _logger = logger;
+            _experimentTimerExpired = false;
+        }
+
+        private void MaxExperimentTimeExpiredEvent(object sender, EventArgs e)
+        {
+            // If at least one model was run, end experiment immediately.
+            // Else, wait for first model to run before experiment is concluded.
+            _experimentTimerExpired = true;
+            if (_history.Any(r => r.RunSucceeded))
+            {
+                _logger.Warning("Allocated time for Experiment of {0} seconds has elapsed with {1} models run. Ending experiment...",
+                    _experimentSettings.MaxExperimentTimeInSeconds, _history.Count());
+                _currentModelMLContext.CancelExecution();
+            }
         }
 
         public IList<TRunDetail> Execute()
         {
-            var stopwatch = Stopwatch.StartNew();
             var iterationResults = new List<TRunDetail>();
+            // Create a timer for the max duration of experiment. When given time has
+            // elapsed, MaxExperimentTimeExpiredEvent is called to interrupt training
+            // of current model. Timer is not used if no experiment time is given, or
+            // is not a positive number.
+            if (_experimentSettings.MaxExperimentTimeInSeconds > 0)
+            {
+                Timer timer = new Timer(_experimentSettings.MaxExperimentTimeInSeconds * 1000);
+                timer.Elapsed += MaxExperimentTimeExpiredEvent;
+                timer.AutoReset = false;
+                timer.Enabled = true;
+            }
```

Comment:
We may need a second timer to check for, and propagate the cancelation signal from the main MLContext to all the new child MLContexts. Then if a user calls `ctx.Cancel()` on the main, the AutoML run will end, just like if the time expired.

Reply:
I don't think we should end all the MLContexts of made experiments. We should always end the MLContext of the training model that is interrupted, which is what the code is currently doing. What if the user would like to use one of the "sub-optimal" models made in the experiment, rather than the model of the best run? If we cancel their respectiveMLContext, they won't be able to use these models. After all, these trained models are all available to use in `experiment.RunDetails.Select(experiment_id).Estimator`. If we don't intend for the user to be able to use these sub-optimal models, we shouldn't even make them available to reach or use.

Reply:
Yes, I think we are saying the same thing. We should only cancel the _one_ active model. 

But currently, if the user, outside of AutoML, cancels the main token, it will not cancel the active training since we don't convey the cancelation from the main token to the one active child token.

Reply:
Got it, thank you. I've added the second timer to check for MLContext's status per 1 second. I'm looking for a better way of checking for this though, perhaps an object listener of some sort where once an object is canceled, the event listener triggers. Any suggestions for those listeners that work on Windows, Linux and MacOS?

Reply:
Looking at the `ICancelable` interface, there's no way to register a delegate which would be auto-executed when the context is canceled:
https://github.com/dotnet/machinelearning/blob/6b2d1a5330bf2f73253edf5895baec367dd329a6/src/Microsoft.ML.Core/Data/IHostEnvironment.cs#L71-L83

I think polling is an ok choice. If `ICancelable` had a way to register a delegate, the code would be a bit more simple (and react on average 0.5s faster). 

@eerhardt may know of a better solution. Does C# have a built-in property watcher (to watch the [IsCanceled property](https://github.com/dotnet/machinelearning/blob/6b2d1a5330bf2f73253edf5895baec367dd329a6/src/Microsoft.ML.Core/Data/IHostEnvironment.cs#L82))? Or perhaps implement a register-delegate function in `ICancelable` or directly in `MLContext`?

Reply:
Typically in .NET we use `CancellationToken` and `CancellationTokenSource`. You can pass a `CancellationToken` around and register an `Action` that will be called when the corresponding `CancellationTokenSource` is cancelled.

Maybe we could use [`CancellationTokenSource.CancelAfter(TimeSpan)`](https://apisof.net/catalog/System.Threading.CancellationTokenSource.CancelAfter(TimeSpan)) instead of the Timer all together?